### ISSUE TITLE
Incorrect handling of nonexistent instances

### ISSUE THREAD

**Issue by jhmartin** [ampedandwired/rancher-reaper#6]
When initially launching the reaper with several 'reconnecting' nodes that have since aged out of AWS:

```
10/31/2016 10:21:15 AMI, [2016-10-31T17:21:15.104529 #5]  INFO -- : Reaping terminated AWS hosts...
10/31/2016 10:21:15 AMW, [2016-10-31T17:21:15.104599 #5]  WARN -- : *** Dry run - no changes will be applied
10/31/2016 10:21:15 AME, [2016-10-31T17:21:15.171469 #5] ERROR -- : undefined method `[]' for nil:NilClass (NoMethodError)
10/31/2016 10:21:15 AM/usr/local/bundle/gems/aws-sdk-resources-2.6.3/lib/aws-sdk-resources/resource.rb:223:in `block in add_data_attribute'
10/31/2016 10:21:15 AM/usr/src/app/reaper.rb:75:in `host_terminated?'
10/31/2016 10:21:15 AM/usr/src/app/reaper.rb:45:in `block in reap_hosts'
10/31/2016 10:21:15 AM/usr/src/app/rancher_api.rb:20:in `yield'
10/31/2016 10:21:15 AM/usr/src/app/rancher_api.rb:20:in `block (3 levels) in get_all'
10/31/2016 10:21:15 AM/usr/src/app/rancher_api.rb:20:in `each'
10/31/2016 10:21:15 AM/usr/src/app/rancher_api.rb:20:in `block (2 levels) in get_all'
10/31/2016 10:21:15 AM/usr/src/app/rancher_api.rb:17:in `loop'
10/31/2016 10:21:15 AM/usr/src/app/rancher_api.rb:17:in `block in get_all'
10/31/2016 10:21:15 AM/usr/src/app/reaper.rb:44:in `each'
10/31/2016 10:21:15 AM/usr/src/app/reaper.rb:44:in `each'
10/31/2016 10:21:15 AM/usr/src/app/reaper.rb:44:in `reap_hosts'
10/31/2016 10:21:15 AM/usr/src/app/reaper.rb:24:in `run'
10/31/2016 10:21:15 AMconfig.ru:24:in `block (2 levels) in <main>'
```

It looks like the API is not triggering the nonexistent instance catch block and is instead triggering a NilClass error.

**Comment by ampedandwired** [comment]
Thanks for the bug report. The code currently handles an `Aws::EC2::Errors::InvalidInstanceIDNotFound` exception, which is what I expected would be thrown if the instance has been purged from AWS. Looks like there might be another scenario though. I'll have a look into it.

**Comment by ampedandwired** [comment]
BTW, in the current implementation if the instance can't be found in the AWS API, it does _not_ get deleted in Rancher. I figured it best to err on the side of caution in cases where it is not definitively clear that the instance has terminated.

If you feel it is useful though, I could add a flag to force deletion in cases like this?

**Comment by Troyhy** [comment]
I have same situation. I have staging environment that is only active from 8-16. In the morning when staging environment goes up, amazon has already purged all info about machine that was serving last afternoon and I get this error. Environment consist only one machine at the moment so reaping cannot happen 16:00 when environment goes down.

It would be helpful to add flag to reap host in this situation.

**Comment by ampedandwired** [comment]
The original issue is actually due to a bug in the AWS SDK: https://github.com/aws/aws-sdk-ruby/issues/1449 - this should be fixed in 0.4.0.

Also, after giving this some thought, I have updated the way rancher-reaper determines if a host is terminated. It now assumes that if the host was not returned by the AWS API, then it is terminated. This should solve the use case mentioned by @Troyhy.
